go/kbfs/search/doc_types.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package search
+
+import (
+	"context"
+	"mime"
+	"net/http"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/blevesearch/bleve/mapping"
+	"github.com/keybase/client/go/kbfs/data"
+	"github.com/keybase/client/go/kbfs/kbfsmd"
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/kbfs/tlf"
+)
+
+const (
+	maxTextToIndex = uint64(10 * 1024 * 1024)
+
+	// Copied from net/http/sniff.go: the algorithm uses at most
+	// sniffLen bytes to make its decision.
+	sniffLen = uint64(512)
+)
+
+type indexedBase struct {
+	Name     string
+	TlfID    tlf.ID
+	Revision kbfsmd.Revision
+	Mtime    time.Time
+}
+
+type indexedTextFile struct {
+	indexedBase
+	Text string
+}
+
+var _ mapping.Classifier = indexedTextFile{}
+
+func (itf indexedTextFile) Type() string {
+	return textFileType
+}
+
+type indexedHTMLFile struct {
+	indexedBase
+	HTML string
+}
+
+var _ mapping.Classifier = indexedHTMLFile{}
+
+func (ihf indexedHTMLFile) Type() string {
+	return htmlFileType
+}
+
+func getContentType(
+	ctx context.Context, config libkbfs.Config, n libkbfs.Node,
+	ei data.EntryInfo) (contentType string, err error) {
+	name := n.GetBasename()
+	contentType = mime.TypeByExtension(filepath.Ext(name.Plaintext()))
```

Comment:
This is dependent on the host environment (`/etc/mime.types`, etc.), and I worry it'd cause confusion like whether a `.js` file is indexed or not (`application/json` vs `text/plain`). Have you considered shipping our own extension list to make this more controllable?

Reply:
@songgao Hmm, I haven't thought about mime types at all.  I just copied what happens in simplefs in terms of figuring out the content type.  I'm not sure why this would need to be different, either way it's per-device behavior controlled by the user, right?  Seems ok to me but I don't consider myself an expert.

Reply:
I'm not an expert either but I think it matters for two reasons:

1. SimpleFS's mime type is for different display mechanisms, and it'd be clear what's going on. This is for deciding whether and how to index a file, which would be a lot subtle to debug.

2. The fact that it's per-device adds confusion -- the index is stored on KBFS, and this means different behavior when it's being indexed from different devices despite of the device it's being used on.

I think `mime.TypeByExtension` makes sense in the context of a HTTP server since it's nice to configure a server with a config file but in general I find it hard to use elsewhere.

Reply:
> The fact that it's per-device adds confusion -- the index is stored on KBFS, and this means different behavior when it's being indexed from different devices despite of the device it's being used on.

The index is not stored on the KBFS servers, if that's what you're implying.  It's stored on the device's local disk, and so far will only be used by that device.  (The code uses KBFSOps and friends as a handy way to encrypt the data, but ultimately it is only stored locally.)  That's why I was saying it's a per-device decision and only affects that device.

That said, future plans might involve one device performing a search on behalf of another device, if that second device doesn't have the TLF synced.  So once that happens, you'd be right that it might be weird if different devices have different behavior.

I can make a ticket for this, but I don't think it makes sense to put it into this PR.  For one thing I don't know how to do it.  But we should probably do it before turning on indexing for anyone in real life, if we suspect it will be a real problem.

Reply:
Oh, I see. I missed that the indices are stored on devices themselves. I don't think having our own TypeByExtension is gonna be very complexity by itself as it's just a map lookup. The complex part is probably to decide which list to use. I'm OK with merging this and deal with that in the future. I see you made a ticket so sounds good to me. Thanks!